sdk/src/Services/S3/Custom/Transfer/Internal/MultipartDownloadManager.cs

```diff
@@ -173,6 +173,41 @@ public Exception DownloadException
             }
         }
 
+        /// <summary>
+        /// Discovers the download strategy (single-part vs multipart) by making an initial GetObject request.
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation token to cancel the discovery operation.</param>
+        /// <returns>
+        /// A <see cref="DownloadDiscoveryResult"/> containing information about the object size, part count,
+        /// and the initial GetObject response.
+        /// </returns>
+        /// <remarks>
+        /// <para><strong>IMPORTANT - HTTP Semaphore Lifecycle:</strong></para>
+        /// <para>
+        /// This method acquires an HTTP concurrency slot from the configured semaphore and downloads Part 1.
+        /// The semaphore slot is <strong>HELD</strong> until <see cref="StartDownloadsAsync"/> completes processing Part 1.
+        /// Callers <strong>MUST</strong> call <see cref="StartDownloadsAsync"/> after this method to release the semaphore.
+        /// Failure to call <see cref="StartDownloadsAsync"/> will cause the semaphore slot to remain held indefinitely,
+        /// potentially blocking other downloads and causing deadlocks.
+        /// </para>
+        /// <para><strong>Concurrency Implications:</strong></para>
+        /// <para>
+        /// With limited HTTP concurrency (e.g., <c>ConcurrentServiceRequests=1</c> for shared throttlers in directory downloads),
+        /// concurrent calls to this method will block until previous downloads complete their full lifecycle
+        /// (discover → start). This is by design to ensure the entire I/O operation (network + disk) is
+        /// within the concurrency limit. For single-slot throttlers, downloads must be processed sequentially:
+        /// complete one download's full lifecycle before starting the next.
+        /// </para>
+        /// <para><strong>Typical Usage Pattern:</strong></para>
+        /// <code>
+        /// var discovery = await manager.DiscoverDownloadStrategyAsync(cancellationToken);
+        /// await manager.StartDownloadsAsync(discovery, progressCallback, cancellationToken);
+        /// await manager.DownloadCompletionTask; // Wait for multipart downloads to finish
+        /// </code>
+        /// </remarks>
+        /// <exception cref="ObjectDisposedException">Thrown if the manager has been disposed.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if discovery has already been performed.</exception>
+        /// <exception cref="OperationCanceledException">Thrown if the operation is cancelled.</exception>
         /// <inheritdoc/>
         public async Task<DownloadDiscoveryResult> DiscoverDownloadStrategyAsync(CancellationToken cancellationToken)
```

Comment:
probably in the future we can refactor this logic such that we dont need a separate DiscoverDownloadStrategyAsyncand we can put everything into `StartDownload`. then at least the acquiring of the http slot and releasing will be in the same function